Various bug fixes

lewisxy opened the pull request:
## Major Changes
- Implement CKA_LABEL, CKA_ID, and some other attributes on public key and private key objects.
- Fix string padding bugs (to comply with the standard)
- Fix segmentation faults when the caller passed null pointers to buffer. Passing null pointers is a defined behavior. The library should return the size of buffer needed.
- Make `FindObjects` more robust by deep copy the filter list in `FindObjectsInit`. This is not an issue for most pkcs11 clients written in C/C++. However, Golang based clients modify the memory used by the filter list between the call to `FindObjects` and `FindObjectsInit`.

## Motivation
I am planning to build a TLS CA on my home lab server (only has TPM 1.2 chip) and the CA software can take advantage of pkcs11 interface for additional security. This library doesn't work out of the box because many parts of the pkcs11 interface are not implemented. Now this library should play well with most pkcs11 clients. I have tested [pkcs11-tool (opensc)][opensc], [python-pkcs11][python-pkcs11], and [step-ca/kms][step-kms].

## (Arbitrary) Design Decisions
The label and ID of the keys are hardcoded strings (see commits). It doesn't really affect anything since those attributes are not used when searching for key objects.

## Unresolved Issues
Minor issue: When a private key no PIN is generated, the PIN (supplied by pkcs11 clients via `C_Login`) are ignored. Not sure this is the intended behavior.
Concurrency: there is no code dealing with concurrent access. Concurrent access behavior is untested.

## Future works (if someone have time...)
- Implement support for multiple keys
- Implement verification (this is not necessary but may be good to have)
- Implement key generation via pkcs11 interface


[opensc]: https://github.com/OpenSC/OpenSC/wiki
[python-pkcs11]: https://github.com/danni/python-pkcs11
[step-kms]: https://github.com/smallstep/step-kms-plugin

lewisxy committed:
fix string bug to comply with standard

lewisxy committed:
implement CKA_ID and CKA_LABEL, fix FindObjects bug

lewisxy committed:
fix sign() to comply with pkcs11 call convention

ThomasHabets commented on src/session.cc:
```diff
@@ -181,6 +193,150 @@ Config::debug_log(const char* fmt, ...) {
 }
 
+bool
+getPublicKeyAttribute(Config& config_, CK_ATTRIBUTE_PTR pAttribute)
```
Trailing underscore is used here for member variables, not function parameters.

(also below)

ThomasHabets commented on src/session.cc:
```diff
@@ -181,6 +193,150 @@ Config::debug_log(const char* fmt, ...) {
+      *(CK_BBOOL *)(pAttribute->pValue) = false;
+    }
+    return true;
+    break;
```
Don't need break after return.
Also below.

ThomasHabets commented on src/session.cc:
```diff
@@ -181,6 +193,150 @@ Config::debug_log(const char* fmt, ...) {
+}
+
+bool
+getPrivateKeyAttribute(Config& config_, CK_ATTRIBUTE_PTR pAttribute)
```
I just made `.debug_log()` a const method, so change to const reference here. (also above)

ThomasHabets commented on src/session.cc:
```diff
@@ -244,10 +439,31 @@ Session::GetAttributeValue(CK_OBJECT_HANDLE hObject,
       config_.debug_log("   Attribute %d: Label (unsupported)", i);
-      if (pTemplate[i].pValue) {
-        *static_cast<char*>(pTemplate[i].pValue) = 'x';
+      if (hObject == 1) {
```
Does 1 and 2 have symbolic values?

lewisxy replied:
These are just my educated guess from the implementation of `object_class` and `FindObjects`. It seems like 1 represent public key and 2 represent private key. It is not documented anywhere. Feel free to document it or make it a bit more clear in the code.

ThomasHabets commented on src/pk11.cc:
```diff
@@ -90,6 +90,14 @@ log_debug(const std::string& msg)
+write_pk11_str(void* dest, char* src, size_t src_size, size_t max_size)
+{
+  memset((char*)dest, 0x20, max_size);
+  memcpy((char*)dest, src, src_size > max_size ? max_size : src_size);
```
Use `std::min(src_size,max_size)`

ThomasHabets commented on src/session.cc:
```diff
@@ -181,6 +193,150 @@ Config::debug_log(const char* fmt, ...) {
   }
 }
 
+bool
```
Make these new functions `static`.

ThomasHabets reviewed (changes requested):
Thanks for the PR! Just some small comments.

If you feel like you can't be bothered, then I can patch it up myself.

ThomasHabets commented:
I'm guessing this fixes https://github.com/ThomasHabets/simple-tpm-pk11/issues/52 for you?

lewisxy committed:
clean up code

lewisxy commented:
I fixed most of the issues. I also refactored the constant strings by putting them in the top of `pk11.cc`. There are a few things I did not change though.
1. Object handle (`hObject`) convention (see comment above)
2. The hardcoded constant strings in `session.cc`. I felt like they might be better to be placed in config file. Right now they are just a local constant. Feel free to make any changes.

ThomasHabets commented:
Merged with some tweaks. Thanks!
